mage_ai/server/api/downloads.py

```diff
@@ -62,3 +68,72 @@ async def get(self, pipeline_uuid, block_uuid, **kwargs):
                     await self.flush()
                     # Sleep for a nanosecond so other handlers can run and avoid blocking
                     await gen.sleep(0.000000001)
+
+
+class ApiResourceDownloadHandler(BaseHandler):
+
+    def get(self, token):
+        try:
+            decoded_payload = jwt.decode(token, JWT_DOWNLOAD_SECRET, algorithms=['HS256'])
+
+            file_name = decoded_payload['file_name']
+            file_list = decoded_payload['file_list']
+            self.ignore_folder_structure = decoded_payload['ignore_folder_structure']
+
+            self.abs_repo_path = os.path.abspath(get_repo_path())
+
+            relative_file_list = list(map(self.relative_path_mapping, file_list))
+
+            try:
+                file_pointer = self.get_file_pointer(file_list, relative_file_list)
+
+                while True:
+                    _buffer = file_pointer.read(4096)
+                    if not _buffer:
+                        break
+                    self.write(_buffer)
+            except Exception as e:
+                self.set_status(400)
+                self.write(f'Error fetching file {file_name}.\n{e}')
+            finally:
+                file_pointer.close()
+
+            self.set_header('Content-Type', 'application/force-download')
+            self.set_header('Content-Disposition', f'attachment; filename={file_name}')
+            self.flush()
+        except jwt.exceptions.ExpiredSignatureError:
+            self.set_status(400)
+            self.write('Download token is expired.')
+        except (jwt.exceptions.InvalidSignatureError, jwt.exceptions.DecodeError):
+            self.set_status(400)
+            self.write('Download token is invalid.')
+        except Exception as e:
+            self.set_status(400)
+            self.write(f'Attepmt at fetching file outside of project folder: {e}')
+
+    # file pointer points to either a singular file or a temporary zip
+    def get_file_pointer(self, file_list, relative_file_list):
+        if len(file_list) == 1:
+            return open(file_list[0])
+        return self.zip_files(file_list, relative_file_list)
+
+    # creates a temporary zip and returns the (open) file pointer
+    def zip_files(self, file_list, relative_file_list):
+        zip_file = tempfile.NamedTemporaryFile(suffix='.zip', mode='w+b')
+        with zipfile.ZipFile(zip_file, 'w') as zipf:
+            for path, relative in zip(file_list, relative_file_list):
+                zipf.write(path, relative)
+        zip_file.seek(0)  # set cursor to start of file to prepare for content extraction
+        return zip_file
+
+    def relative_path_mapping(self, path):
+        abs_path = os.path.abspath(path)
+        common_ground = os.path.commonpath([self.abs_repo_path, abs_path])
+
+        # trying to access files outside of the project folder
+        if common_ground != self.abs_repo_path:
+            raise Exception(abs_path)
```

Comment:
Can we make this Exception more specific? https://docs.python.org/3/library/exceptions.html#bltin-exceptions.. I am thinking PermissionError makes a bit more sense

Reply:
I think **PremissionError** makes it sound like there are certain permissions that would allow you to download files outside the scope of the project, which there are none. As such I replaced it with [**ValueError**](https://docs.python.org/3/library/exceptions.html#ValueError) as I think it makes the most sense.